clusterloader2/pkg/measurement/common/service_creation_latency.go

```diff
@@ -0,0 +1,317 @@
+
+			}()
+			klog.Infof("%s: %d created, %d ipAssigned, %d reachable", s, created, ipAssigned, reachable)
+			if created == reachable {
+				return nil
+			}
+		}
+	}
+}
+
+func (s *serviceCreationLatencyMeasurement) gather() ([]measurement.Summary, error) {
+	klog.Infof("%s: gathering service created latency measurement...", s)
+	if !s.isRunning {
+		return nil, fmt.Errorf("metric %s has not been started", s)
+	}
+
+	serviceCreationLatency := s.creationTimes.CalculateTransitionsLatency(map[string]measurementutil.Transition{
+		"create_to_assigned": {
+			From: creatingPhase,
+			To:   ipAssigningPhase,
+		},
+		"assigned_to_available": {
+			From: ipAssigningPhase,
+			To:   reachabilityPhase,
+		},
+		"create_to_available": {
+			From: creatingPhase,
+			To:   reachabilityPhase,
+		},
+	})
+
+	content, err := util.PrettyPrintJSON(measurementutil.LatencyMapToPerfData(serviceCreationLatency))
+	if err != nil {
+		return nil, err
+	}
+	summary := measurement.CreateSummary(serviceCreationLatencyName, "json", content)
+	return []measurement.Summary{summary}, err
+}
+
+func (s *serviceCreationLatencyMeasurement) handleObject(oldObj, newObj interface{}) {
+	var oldService *corev1.Service
+	var newService *corev1.Service
+	var ok bool
+	oldService, ok = oldObj.(*corev1.Service)
+	if oldObj != nil && !ok {
+		klog.Errorf("%s: uncastable old object: %v", s, oldObj)
+		return
+	}
+	newService, ok = newObj.(*corev1.Service)
+	if newObj != nil && !ok {
+		klog.Errorf("%s: uncastable new object: %v", s, newObj)
+		return
+	}
+	isEqual := oldService != nil &&
+		newService != nil &&
+		equality.Semantic.DeepEqual(oldService.Spec, newService.Spec) &&
+		equality.Semantic.DeepEqual(oldService.Status, newService.Status)
+	if isEqual {
+		return
+	}
+
+	if !s.isRunning {
+		return
+	}
+	if newObj == nil {
+		if err := s.deleteObject(oldService); err != nil {
+			klog.Errorf("%s: delete checker error: %v", s, err)
+		}
+		return
+	}
+	if err := s.updateObject(newService); err != nil {
+		klog.Errorf("%s: create checker error: %v", s, err)
+	}
+}
+
+func (s *serviceCreationLatencyMeasurement) deleteObject(svc *corev1.Service) error {
+	key, err := runtimeobjects.CreateMetaNamespaceKey(svc)
+	if err != nil {
+		return fmt.Errorf("meta key created error: %v", err)
+	}
+	s.pingCheckers.DeleteAndStop(key)
+	return nil
+}
+
+func (s *serviceCreationLatencyMeasurement) updateObject(svc *corev1.Service) error {
+	if len(svc.Spec.Ports) < 1 {
+		return nil
+	}
+	key, err := runtimeobjects.CreateMetaNamespaceKey(svc)
+	if err != nil {
+		return fmt.Errorf("meta key created error: %v", err)
+	}
+	if _, exists := s.creationTimes.Get(key, creatingPhase); !exists {
+		s.creationTimes.Set(key, creatingPhase, time.Now())
+	}
+	if len(svc.Status.LoadBalancer.Ingress) < 1 {
+		return nil
+	}
+	if _, exists := s.creationTimes.Get(key, ipAssigningPhase); exists {
+		return nil
+	}
+	s.creationTimes.Set(key, ipAssigningPhase, time.Now())
+	klog.Infof("%s: %s ipAssigned", s, key)
```

Comment:
Do we really want to log it?

Reply:
We are doing similar thing when waiting for pods. I found it convenient when debugging, so I want to have similar logic here.

Reply:
We're logging every change state for every pod? I highly doubt.

I looked into wait_for_pods, wait_for_controlled_pods and pod_startup_slo and in neither of those you log it.
If you really log it, please point me to it.